[sources] Respect Debezium transaction boundaries

cjubb39 opened the pull request:
Read in Debezium transaction data alongside the data so that we can keep everything from the same transaction at the same timestamp.  We do this with the following way:
- Create a source that contains the transaction data in the format described here: https://debezium.io/documentation/reference/connectors/postgresql.html#postgresql-transaction-metadata
- Create a second source using `WITH tx_metadata = ??`

Internally this second source creates a binary timely operator that reads in both streams of data and matches them up.  On each invocation this operator advances to the `meet` of the tx metadata and data frontiers when there's no outstanding transaction or data to match up.  If either of these are outstanding, we should not advance the frontier.

It may be possible to come up with a tighter bound but I'd like to avoid diving into that too much in this PR -- but please let me know if you think we should tightening it and I'll make a follow-up item

### Motivation
Fixes #8693
Fixes #8695
Fixes #6296
Fixes #6555

### Tips for reviewer
Files by type of changes (hopefully this makes what's going on more clear)
- `src/dataflow/src/render/debezium.rs`: this is where we actually do the merging of the tx metadata and the data
- `src/dataflow/src/render/sources.rs`: this is where we re-create the source that contains the tx metadata
- `src/sql/src/plan/statement/ddl.rs`: this is where we validate the shape of the tx metadata source (including that it has the proper columns) and setting up the dependencies for the data source (e.g. we don't want to allow the metadata source to be dropped if the data source is still alive)
- `test/testdrive/kafka-avro-debezium-transaction.td` new tests!

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

- `tx_metadata` / `tx_metadata_collection_name` are now accepted as options for debezium sources so that they're able to be transactional.  This lives behind an experimental flag.  Multiple data sources can share a single transaction metadata source

guswynn commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,210 @@
+
+> CREATE MATERIALIZED SOURCE data_schema_inline
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-${testdrive.seed}'
+  WITH(tx_metadata = data_txdata)
```
why do we point to a source instead of just the schema? then we could build the source exactly as expected?

cjubb39 replied:
I'm going down this path for a couple reasons:
- I think it makes the structure more intuitive to a user
- Make more explicit some internal complexity: if we inlined the kafka info, then we'd likely make something that looks like a source but we wouldn't want to show to the user, etc
- Allow for explicit re-use of the tx topic since we do want this to be able to work cross-topic.  As you noted above, we don't currently do that, but we'll want to in the future!

guswynn commented on src/dataflow/src/render/sources.rs:
```diff
@@ -435,11 +435,57 @@ where
+                                                dataflow_debug_name,
+                                                dataflow_id,
+                                                as_of_frontier,
+                                                SourceInstanceDesc {
```
is this the first example of making multiple source instances when we make a source instance?
Also, does using the same as_of_frontier make sense?

cjubb39 replied:
Not sure if it's the first but that is what's happening here yeah.  Will def be a TODO to fix in future work!

guswynn commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +129,259 @@ where
+    impl<T: std::fmt::Debug> Iterator for VdIterator<T> {
+        type Item = T;
+        fn next(&mut self) -> Option<T> {
+            eprintln!("PROVIDING {:?}", self.0.borrow_mut().get(0));
```
for `debug` messages, you might prefer the `dbg!` macros, which lets you do stuff like
```
return dbg!(self.0.borrow_mut().pop_front())
```
and it passes the value through, and prints the expression that caused it and line number

cjubb39 replied:
good tip!

cjubb39 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -602,6 +604,76 @@ pub fn plan_create_source(
+                    // TODO(chae): validate shape of tx_metadata item: maybe including full dbz dedup?
+                    let tx_metadata = match with_options.remove("tx_metadata") {
+                        Some(Value::String(src)) => {
+                            // XXX(chae): this seems like I'm breaking a boundary by accessing scx.catalog
```
@jkosh44 do you mind taking a quick look at this?  Not sure if I'm doing this the right way -- or if there's a better way to convert `src` into `item`

jkosh44 replied:
In terms of looking up an item in the catalog, getting the full_name, and getting the description it all looks correct to me. However it make me a little nervous to see you hard-code `"materialize.public"` as the database and schema name. I don't really know the context of this PR, but it is possible to delete the `materialize` database and the `public` schema and things are supposed to continue to work as normal.

jkosh44 replied:
If you don't know what database and schema `src` lives in, then just use `UnresolvedObjectName::unqualified` and the catalog has a search process to look for the item in various databases and schemas. Include `materialize.public` if it exists.

cjubb39 replied:
Yeah I didn't like `materialize.public` either -- the issue I had with doing `UnresolvedObjectNmae::unqualified(&src)` is that when restarting, I get an error that `$whatever_was_in_src` is an unknown catalog item -- with the error stemming from here

Do you think this might be a good case for purification to do the resolution?

jkosh44 replied:
tbh honest I don't exactly know what purification is. Is `src` already a qualified string? If so you might have to split it on periods. If you give me a way to re-create that error I can take a look at it. It's odd that it wouldn't be able to find it because `resolve_item` will look in `materialize.public` if it exists.

cjubb39 replied:
We chatted in slack and I've changed the approach slightly to make sure we can handle the initial `CREATE SOURCE` call as well as its recreation after restarts:
- let `resolve_item` do it's thing on the initial call
- use normalize::create_statement to persist this normalization (often prepending `materialize.public.`) so that it's used on subsequent restarts

This normalization mirrors how we normalize the name of the source itself

jkosh44 commented on src/sql/src/normalize.rs:
```diff
@@ -306,6 +306,14 @@ pub fn create_statement(
+            for opt in with_options.iter_mut() {
+                if let SqlOption::ObjectName { name, object_name } = opt {
+                    if ident_ref(name) == "tx_metadata" {
+                        *object_name = allocate_name(object_name)?;
```
So I think `allocate_name` will probably result in the right name most of the time, but I don't think it's exactly correct. Since `object_name` is an existing object, we should probably be looking for it in the catalog instead of assuming it's database and schema which `allocate_name` sort of does.

Instead you can resolve the object with with `scx.resolve_item` and get the fully qualified name that way.

Another option, which I think is probably better but relies on `object_name` always being a valid catalog object, is to apply the following patch (I don't know how to attach a patch on GH):
```
From 95b3de8247aff77ec9e8af728234c26b76e6a561 Mon Sep 17 00:00:00 2001
From: Joseph Koshakow <koshy44@gmail.com>
Date: Wed, 30 Mar 2022 17:23:06 -0400
Subject: [PATCH] Auto resolve

---
 src/sql-parser/src/ast/defs/statement.rs | 2 +-
 src/sql-parser/src/parser.rs             | 2 +-
 src/sql/src/normalize.rs                 | 8 --------
 3 files changed, 2 insertions(+), 10 deletions(-)

diff --git a/src/sql-parser/src/ast/defs/statement.rs b/src/sql-parser/src/ast/defs/statement.rs
index cbe666193..d56d0c684 100644
--- a/src/sql-parser/src/ast/defs/statement.rs
+++ b/src/sql-parser/src/ast/defs/statement.rs
@@ -1698,7 +1698,7 @@ pub enum SqlOption<T: AstInfo> {
     },
     ObjectName {
         name: Ident,
-        object_name: UnresolvedObjectName,
+        object_name: T::ObjectName,
     },
     DataType {
         name: Ident,
diff --git a/src/sql-parser/src/parser.rs b/src/sql-parser/src/parser.rs
index 6351aee1c..66cf272e7 100644
--- a/src/sql-parser/src/parser.rs
+++ b/src/sql-parser/src/parser.rs
@@ -2849,7 +2849,7 @@ impl<'a> Parser<'a> {
             SqlOption::Value { name, value }
         } else {
             self.prev_token();
-            if let Ok(object_name) = self.parse_object_name() {
+            if let Ok(object_name) = self.parse_raw_name() {
                 SqlOption::ObjectName { name, object_name }
             } else {
                 self.expected(self.peek_pos(), "option value", token)?
diff --git a/src/sql/src/normalize.rs b/src/sql/src/normalize.rs
index 4df95066f..cc2eda18b 100644
--- a/src/sql/src/normalize.rs
+++ b/src/sql/src/normalize.rs
@@ -306,14 +306,6 @@ pub fn create_statement(
             *name = allocate_name(name)?;
             *if_not_exists = false;
             *materialized = false;
-
-            for opt in with_options.iter_mut() {
-                if let SqlOption::ObjectName { name, object_name } = opt {
-                    if ident_ref(name) == "tx_metadata" {
-                        *object_name = allocate_name(object_name)?;
-                    }
-                }
-            }
         }
 
         Statement::CreateTable(CreateTableStatement {
```

That will automatically resolve and full qualify `object_name` as part of planning.

cjubb39 replied:
Thanks for writing up the patch! I'm not sure if that generalizes / I don't want to wrap that change into this PR but it could be good to look into for the future -- and I do like the idea of not special casing tx_metadata but alas

I'll make a change to take the `resolve_item` route as you suggest

cjubb39 commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,345 @@
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-data-${testdrive.seed}'
+  WITH(tx_metadata = data_txdata)
+  FORMAT AVRO USING SCHEMA '${schema}'
+  ENVELOPE DEBEZIUM
```
Thoughts on where the new option lives syntactically?  Is it better suited being something like `ENVELOPE DEBEZIUM TRANSACTION_METADATA $name_of_source`?

guswynn replied:
I think explicit keywords like this are better than `WITH`!

jkosh44 reviewed (approved):
LGTM from a SQL point of view (I did not look at anything in the `dataflow` create).

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+        } => panic!("render_tx should be called with a dedup mode"),
+    };
+
+    // XXX(chae): Should we try to avoid moving everything to one worker?
```
exchanging based on transaction_id should work I think

petrosagg replied:
Hm, I'm not sure I follow. Why can't we write something like this? (with unwraps turned into some sensible default

``rust
Exchange::new(move |decode_result| {
    decode_result
        .value
        .unwrap()
        .iter()
        .nth(transaction_idx)
        .unwrap()
        .unwrap_list()
        .iter()
        .nth(tx_metadata_description.data_transaction_id_idx)
        .unwrap()
        .unwrap_str()
        .hashed()
)
```

cjubb39 replied:
oh hmm good point. I think this should work (with, yeah, some defaults) because if, e.g. decode_result.value.is_none(), we don't have any hope of matching the transaction anyway so it doesn't matter which worker it ends up on!

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                                let event_count = match event_count {
+                                    Some(c) => c,
+                                    None => {
+                                        // XXX(chae): maybe a panic??
```
Yes I think this is the correct behaviour. We should probably flip a flag so that the operator permanently enters an errored state in which the only thing it does is empty its inputs and drop them on the floor

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                    move |tx_metadata_input, data_input, output| {
+                        while let Some((tx_metadata_cap, refmut_data)) = tx_metadata_input.next() {
+                            refmut_data.swap(&mut tx_data);
+                            let tx_metadata_cap = Rc::new(tx_metadata_cap.retain());
```
You can drop the `Rc<_>` here:

```diff
diff --git a/src/dataflow/src/render/debezium.rs b/src/dataflow/src/render/debezium.rs
index 2a53ec06a..72ca51e28 100644
--- a/src/dataflow/src/render/debezium.rs
+++ b/src/dataflow/src/render/debezium.rs
@@ -203,7 +203,7 @@ where
                     move |tx_metadata_input, data_input, output| {
                         while let Some((tx_metadata_cap, refmut_data)) = tx_metadata_input.next() {
                             refmut_data.swap(&mut tx_data);
-                            let tx_metadata_cap = Rc::new(tx_metadata_cap.retain());
+                            let tx_metadata_cap = tx_metadata_cap.retain();
                             for (row, time, diff) in tx_data.drain(..) {
                                 assert_eq!(diff, 1);
                                 let mut i = row.iter();
@@ -224,7 +224,7 @@ where
                                     Some(c) => c,
                                     None => {
                                         // XXX(chae): maybe a panic??
-                                        output.session(&*tx_metadata_cap).give((
+                                        output.session(&tx_metadata_cap).give((
                                             Err(DataflowError::EvalError(EvalError::Internal(
                                                 String::from("Need events_remaining in END record"),
                                             ))),
@@ -238,7 +238,7 @@ where
                                     None => {
                                         tx_event_count.insert(tx_id.clone(), event_count);
                                         tx_cap_map
-                                            .insert(tx_id.clone(), Rc::clone(&tx_metadata_cap));
+                                            .insert(tx_id.clone(), tx_metadata_cap.clone());
                                     }
                                     Some(val) if val == time => {},
                                     Some(val) => panic!("unexpected mismatch in duplicate END record for {:?}: {:?} vs {:?}", tx_id, time, val),
@@ -312,7 +312,7 @@ where
                                             // XXX(chae): is it worth trying to use `tx_cap` if it exists or should just use `binary_cap`?
                                             // If we would dedup the message below, we might not have a cap
                                             match tx_cap_map.get(tx_id) {
-                                                Some(c) => output.session(&**c).give((
+                                                Some(c) => output.session(c).give((
                                                     Err(err),
                                                     *c.time(),
                                                     1,
@@ -336,7 +336,7 @@ where
                                     Entry::Vacant(_) => panic!("Must have cap if using record"),
                                 };

-                                let mut session = output.session(&**tx_cap_entry.get());
+                                let mut session = output.session(tx_cap_entry.get());
                                 match value.iter().nth(before_idx).unwrap() {
                                     Datum::List(l) => {
                                         session.give((Ok(Row::pack(&l)), tx_time, -1));
```

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                            let tx_metadata_cap = Rc::new(tx_metadata_cap.retain());
+                            for (row, time, diff) in tx_data.drain(..) {
+                                assert_eq!(diff, 1);
+                                let mut i = row.iter();
```
nit: a better named variable would be nice, maybe `datums`or smth

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                                    match res {
+                                        Ok(b) => b,
+                                        Err(err) => {
+                                            // XXX(chae): is it worth trying to use `tx_cap` if it exists or should just use `binary_cap`?
```
I don't think we should be using the initial `binary_cap` anywhere in this operator because all the data that it is producing are a reaction on its inputs, which already come with capabilities. So we should just drop it on the floor during initialization (which happens automatically if you don't touch it).

I think our strategy should be that all data events are emitted with the corresponding `tx_cap` and all errors are emitted with the corresponding `data_cap`.

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                        }
+
+                        if !data_buffer.is_empty() || !tx_cap_map.is_empty() {
+                            activator.activate_after(Duration::from_secs(1));
```
We don't need to manually activate the operator here because we won't make progress anyway by reprocessing the state. The only way to make forward progress is for new data to arrive, which will cause timely to schedule us anyway.

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+                            // then `take`ing it if `meet` is the empty Antichain. But then we'd have a
+                            // bunch of unwraps everywhere.
+                            let meet_u64 = meet.into_iter().next().unwrap_or(u64::MAX);
+                            let _ = binary_cap.try_downgrade(&meet_u64);
```
Following from the comment further up, we should just leave `binary_cap` alone to get dropped during initialization and work solely with the capabilities of the inputs

petrosagg commented on src/dataflow/src/render/sources.rs:
```diff
@@ -438,11 +436,48 @@ where
+                                            .expect("bad tx metadata spec")
+                                            .clone();
+                                        let ((tx_source_ok, tx_source_err), tx_token) =
+                                            import_source(
```
There is probably some logic missing here where we reuse an arrangement if there is one, right?

petrosagg commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -602,6 +605,76 @@ pub fn plan_create_source(
+                                .context("tx_metadata catalog item doesn't have a desc")?
+                                .iter();
+
+                            match desc_iter.next() {
```
This validation is a bit different than the debezium one. In the debezium case we don't care about the column indexes, we just care that the columns have the correct name and type, regardless of their index. I'd vote for being consistent in what users are expected to offer

petrosagg commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -602,6 +605,76 @@ pub fn plan_create_source(
+                            }
+
+                            depends_on.push(item.id());
+                            depends_on.extend(item.uses());
```
TIL: `depends_on` is transitive!

petrosagg commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -602,6 +605,76 @@ pub fn plan_create_source(
+
+                            Some(item.id())
+                        }
+                        Some(v) => bail!("tx_metadata must be a Object found {:?}", v),
```
missing period `.`, I think

petrosagg reviewed (commented):
Did a first pass of this, exciting!

guswynn commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +124,273 @@ where
+            ..
+        } => (
+            *transaction_idx,
+            tx_id_idx.expect("render_tx should only be called when there's a transaction"),
```
do we validate this somehow before? 1 bad message from debezium could cause a panic, right?

guswynn commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -602,6 +604,71 @@ pub fn plan_create_source(
                     UnplannedSourceEnvelope::Upsert(UpsertStyle::Debezium { after_idx })
                 }
                 DbzMode::Plain => {
+                    // XXX(chae): should this be for upsert too?
```
upsert + dbz is handled mostly separately from the rest of dbz now: https://github.com/MaterializeInc/materialize/pull/11306/files#diff-c840a7fbb6a642d6cbf71dddd92ec1b6dfa86310022ead5673e8fa509ba73dc4L602, its a mode under `SourceEnvelope::Upsert`

we could probably make it work with dbz transactions, but I assume we would need to have the tx-timestamp merging logic manage the stream that is inputted into the upsert operator?

cjubb39 commented:
Outstanding issues to address (let me know if I've missed -- and sorry)
- error handling (I've been discussing this with petros)
- exact syntax for identifying the transaction metadata source
- identifying the columns by name in the transaction meta data source (I'm working on this)

I wanted to push up a version real quick so that Petros and I could refer to something concrete when discussing error handling -- since I did address a few CR comments

cjubb39 commented:
Only things outstanding for now:
1) Exact syntax.  I'd like a concrete suggestion if anyone wants me to do something different than a `WITH` option
2) Updating the docs.  I think this is okay as a follow-up.  It might even make sense to have this happen once I'm done with this / immediate follow-ups like making it work for a metadata topic that encompasses multiple data topics
3) Error handling.  I have a doc [here](https://docs.google.com/document/d/1kJuDuja_5UbzStL8-2uYR6bQsVQDXqmZG4Wf5q3qYfc/view) that describes some approaches but overall I think it's tricky and hinges on what we consider an acceptable/desirable user behavior.  The current handling (i.e. not handling errors and just continuing to emit data points) provides a definite source so it's a correct implementation.  Pushing those changes to a second PR will allow easier parallel work with supporting multiple data topics.

All that to say: I think we can merge this change in (finally!) and we'll have a working implementation even if there's a possibility of large memory usage in the error case -- because I'll immediately start work on addressing that separately.  What do you think?

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +121,265 @@ where
+                            };
+
+                            if should_use {
+                                let tx_cap_entry = match tx_cap_map.entry(tx_id.clone()) {
```
this looks like a `tx_cap_map.get(&tx_id).expect("Must have cap if using record")`

cjubb39 replied:
I used the Entry API here to make the removal below more ergonomic -- so we only panic in one spot.  I was attempting to minimize explicit panics -- especially given how many we need to have given the data structures we're using to track things.  I think it's easier to reason about correctness this way.  What do you think?

petrosagg replied:
I missed the removal further down, sorry. I retract my suggestion, that looks fine

petrosagg commented on src/dataflow/src/render/debezium.rs:
```diff
@@ -118,6 +121,265 @@ where
+                                        panic!("type error: expected record, found {:?}", d)
+                                    }
+                                }
+                                match tx_event_count.entry(tx_id.clone()) {
```
same here, you can use `get_mut(&tx_id).expect(..)`

petrosagg replied:
(well, here you need `.unwrap_or_else(|| panic!(..))` but it's still simpler than doing through the entry API

cjubb39 replied:
I was going for a similar thing here: it seemed to minimize panics unwraps when we only sometimes remove the entry after reading it.  Is there a magical rust way to do that if we use `get_mut`?

petrosagg replied:
You're right, I didn't read it carefully enough, sorry. I'll resolve both of these comments

cjubb39 commented:
sorry to dump a bunch of info at you but allowing multiple data streams to use a single transaction stream was a small change here that I'm happy to put on this branch and merge in if you think it makes sense: https://github.com/cjubb39/materialize/pull/1   It's written such that this PR is the base of that one

philip-stoev commented:
@cjubb39 . We have a dedicated test for debezium that uses actual database products to generate transactions. Would it be possible to add test cases test/debezium for the functionality that is enabled by this PR. Testing this using test/testdrive is does not provide complete realism.

At the very least, the test case from  https://github.com/MaterializeInc/materialize/issues/6555 needs to be included.

cjubb39 commented:
Okay.  I have a few notes about changes and should be good for a big review from @petrosagg :

### Testing
I have adjusted the tests philip referenced in #6555.  I'm not sure there were more tests to write (other than testdrive) -- at least at the moment.  I've added proper `test/debezium` testing for postgres and mysql backed dbz but passed on sql-server for now because I can't get it to emit an `END` record for the most recent transaction lol.  I did make the postgres test smaller as 1M records was taking a very long time.  10k seemed sufficient. (100k also took a few minutes to process the `UPDATE` statement)

### New Features since last updated
- Multiple topics: In order to get the pre-existing testing to work, I merged in support for multiple data collections that use the same transaction metadata.  I already had it written up and it wasn't that complicated so I just merged.
- Support for data before transactions started: Messages that have a null transaction metadata field are just processed straight through, passed on unchanged at the same capability.  Assuming the input data stream is definite, the output will also be definite.  This is tested explicitly in the `test/debezium/mysql` tests.

### Syntax
I've put this feature behind the experimental flag so I'm not super worried that we're still using a `WITH` option.  I also needed to add a second `WITH` option so you're able to specify the name of the collection as provided in the `data_collections` field of the transaction metadata messages.

### Error handling
Chatted in person with Petros.  Doing this more intelligently than just not matching the data record and holding back the frontier will be pushed down the road.  Doing it correctly while keeping the output collection still definite will take some work / will be subtle to get right.  Plus our concerns about memory usage are theoretical and I wanted to keep moving

### Docs
It's experimental so I was not going to add docs

cjubb39 commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+        } => panic!("render_tx should be called with a dedup mode"),
+    };
+
+    // XXX(chae): Should we try to avoid moving everything to one worker?
```
Petros: I'll fix this in a follow-up.  Will make a GH issue

cjubb39 commented on src/dataflow/src/storage/render/sources.rs:
```diff
@@ -438,11 +438,49 @@ where
+                                            // N.B. tx_id is validated when constructing dbz_envelope
+                                            .expect("bad tx metadata spec")
+                                            .clone();
+                                        // TODO(chae): reuse the existing arrangement if it exists
```
Petros: I'll make a GH issue for this

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                                    let tx_id = l.iter().nth(tx_metadata_description.data_transaction_id_idx).unwrap().unwrap_str().to_owned();
+                                    let tx_time: Timestamp = match tx_mapping.get(&tx_id) {
+                                        Some(time) => *time,
+                                        None => break,
```
This is the one control flow path where the element stays in the deque and is not popped so I would flip the logic around and pop the element as part of the while loop and only put it back here right before breaking.

Here is a proposed diff:

```diff
diff --git a/src/dataflow/src/storage/render/debezium.rs b/src/dataflow/src/storage/render/debezium.rs
index bc1cd13e6..3f8bad83b 100644
--- a/src/dataflow/src/storage/render/debezium.rs
+++ b/src/dataflow/src/storage/render/debezium.rs
@@ -269,8 +269,8 @@ where
                             let data_cap = data_cap.retain();
                             data_buffer.extend(data.drain(..).map(|r| (r, data_cap.clone())));
                         }
-                        while let Some((result, data_cap)) = data_buffer.get(0).cloned() {
-                            let key = match result.key.transpose() {
+                        while let Some((result, data_cap)) = data_buffer.pop_front() {
+                            let key = match result.key.clone().transpose() {
                                 Ok(key) => key,
                                 Err(err) => {
                                     output.session(&data_cap).give((
@@ -278,33 +278,35 @@ where
                                         *data_cap.time(),
                                         1,
                                     ));
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                             };
-                            let value = match result.value {
-                                Some(Ok(value)) => value,
+                            let value = match result.value.clone() {
+                                Some(Ok(value)) => value.clone(),
                                 Some(Err(err)) => {
                                     output.session(&data_cap).give((
                                         Err(err.into()),
                                         *data_cap.time(),
                                         1,
                                     ));
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                                 None => {
-                                    let _ = data_buffer.pop_front();
                                     continue;
                                 }
                             };

                             let tx_id = match value.iter().nth(transaction_idx).unwrap() {
                                 Datum::List(l) => {
-                                    let tx_id = l.iter().nth(tx_metadata_description.data_transaction_id_idx).unwrap().unwrap_str().to_owned();
+                                    let tx_id = l.iter().nth(data_tx_id_idx).unwrap().unwrap_str().to_owned();
                                     let tx_time: Timestamp = match tx_mapping.get(&tx_id) {
                                         Some(time) => *time,
-                                        None => break,
+                                        None => {
+                                            // We don't yet have a time for this piece of data so
+                                            // bail until we get more transaction metadata
+                                            data_buffer.push_front((result, data_cap));
+                                            break;
+                                        }
                                     };
                                     Some((tx_id, tx_time))
                                 },
@@ -312,9 +314,6 @@ where
                                 _ => panic!("Previously validated to be nullable list"),
                             };

-                            // We're committed to processing it if we have a tx_time
-                            let _ = data_buffer.pop_front();
-
                             let partition_dedup = dedup_state
                                 .entry(result.partition.clone())
                                 .or_insert_with(|| {
```

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                            // We're committed to processing it if we have a tx_time
+                            let _ = data_buffer.pop_front();
+
+                            let partition_dedup = dedup_state
```
Can you leave a TODO comment here about re-using the dedup operator we already have?

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                                }
+                            };
+
+                            let tx_id = match value.iter().nth(transaction_idx).unwrap() {
```
This variable is named `tx_id` but it actually contains both `tx_id` and `tx_time`

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                // Hold onto a capability for each `transaction_id`.  This represents the time at which we'll emit
+                // matched data rows.  This will be dropped when we've matched the indicated number of events.
+                let mut tx_cap_map = HashMap::new();
+                let mut tx_event_count = HashMap::new();
```
We should probably keep a single map that maps a transaction id to a `(time, cap, event_count)` tuple. Below in the logic these three maps are populated at the same time and the existence of a value in one is used as evidence for the existence of a value in the other ones (the places where we panic on Vacant entries). If this was a single map where all the information is inserted and extracted in one place there would be no possibility for misuse.

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                                    continue;
+                                }
+
+                                match tx_mapping.insert(tx_id.clone(), time) {
```
Hm, I don't see any place where we empty this map. I think we're a `.remove()` down below where you clear the other maps. But see the comment above regarding having a single map

petrosagg commented on src/dataflow/src/storage/render/debezium.rs:
```diff
@@ -118,6 +121,304 @@ where
+                                        Entry::Vacant(_) => panic!("need event count for tx_id {:?}", tx_id),
+                                    }
+                                },
+                                (true, None) => {
```
Hm, shouldn't this be an error or ignored? In which case do we care about data that cannot be matched to a transaction?

petrosagg commented on src/dataflow-types/src/types.rs:
```diff
@@ -899,6 +900,18 @@ pub mod sources {
 
+    #[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+    pub struct DebeziumTransactionMetadata {
+        pub tx_metadata_global_id: GlobalId,
```
NIT: I would drop the `tx_` prefix since it is implied by the type name and will make them shorter

petrosagg commented on src/dataflow-types/src/types.rs:
```diff
@@ -955,6 +982,8 @@ pub mod sources {
         pub transaction_idx: usize,
         /// The record index of the `transaction.total_order` field
         pub total_order_idx: usize,
+        /// The record index of the `transaction.id` field
```
This contains a lot more than the `transaction.id` field, right?

petrosagg commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,365 @@
+  WITH(tx_metadata = data_sink, tx_metadata_collection_name = 'testdrive-data-${testdrive.seed}')
+  FORMAT AVRO USING SCHEMA '${schema}'
+  ENVELOPE DEBEZIUM
+contains:tx_metadata catalog item doesn't have a desc
```
Hm, should we give a more descriptive error like "tx_metadata must refer to a source, table, or view"?

petrosagg commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,365 @@
+# Test reading from the source when tx and data don't match
+#
+# Sleep to make sure we get a different RT timestamp from previous messages
+$ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration=5s
```
I think we remove these sleeps and replace them with this which will wait exactly the right amount of time for the RT timestamp to progress:

```sql
> BEGIN
> DECLARE CURSOR c FOR TAIL data_schema_inline WITH (snapshot = false, progress = true)
> FETCH 1 c
mz_timestamp  mz_progressed  mz_diff  a       b
-------------------------------------------------
<timestamp>   true           <null>   <null>  <null>
> COMMIT
```

I believe testdrive runs with timestamp frequency set to 100ms

petrosagg commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,365 @@
+
+$ set-regex match=\d{13} replacement=<TIMESTAMP>
+
+$ kafka-verify format=avro sink=materialize.public.data_sink sort-messages=true
```
So one of the things that we haven't tested but I think it's worth adding is that the transaction boundaries are maintained and reflected by our timestamp assignment. It's a bit hard to do entirely within materialize because afaik the only command that exposes the underlying timestamps is TAIL which cannot be used in more complex queries and we don't really know the assigned timestamp to predict the output.

One thing we can do however is create a kafka source pointed to the data_sink that mz produces and then for example run a `SELECT COUNT(*) AS event_count GROUP BY transaction.id ORDER BY transaction.id` query that will verify that there were 3 distinct transactions with the expected number of transactions.

cjubb39 replied:
clever!  will update

petrosagg commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,365 @@
+8  9
+2  7
+
+$ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration=2s
```
See comment above for waiting exactly the right amount of time

petrosagg reviewed (approved):
I think this is good to merge! I left some comments on the tests but feel free to merge as soon as you think those are addressed :tada: :tada: :tada: :tada:

cjubb39 commented on test/testdrive/kafka-avro-debezium-transaction.td:
```diff
@@ -0,0 +1,365 @@
+#
+# Create a source using a file schema
+#
+! CREATE MATERIALIZED SOURCE data_schema_file
```
good point.  it's a duplicate.  I'll remove

cjubb39 commented:
I'm gonna rebase to try to avoid conflicts, then it'll be time to bit the big fancy button and merge!
